### PR TITLE
[SYCL][DOC] Add a document covering linked allocation commands

### PR DESCRIPTION
**Pull request by sergey-semenov**

### CODE REVIEW

**Review by alexbatashev — CHANGES_REQUESTED** [review]
Could you please add this to index.rst?

File: sycl/doc/LinkedAllocations.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
+## Brief overview of allocations for memory objects
+
+A SYCL memory object (`buffer`/`image`) can be accessed in multiple contexts
+throughout its lifetime. Since this is dynamic information that is unknown
+during memory object construction, no allocation takes place at that point.
+Instead, memory is allocated in each context whenever the SYCL memory object
+is first accessed there:
+
+```C++
```
Comment (alexbatashev) [inline sycl/doc/LinkedAllocations.md:9]: Also current version of Sphinx doesn't process these hints correctly (github also has hard times rendering my suggestion correctly, but I propose to remove `C++`).
```suggestion
```<just_quotes>
```

**Review by alexbatashev — DISMISSED** [review]
LGTM now

**Review by pvchupin — DISMISSED** [review]
Would be good to have more people review.

File: sycl/doc/LinkedAllocations.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+shared between the two (the underlying backend is still free to ignore that
+request and allocate additional memory if needed). The difference in handling
+linked and unlinked allocations is summarized in the table below.
+
+|   | Unlinked | Linked |
+| - | -------- | ------ |
+| Native memory object creation | Created with COPY_HOST_PTR if a host pointer is available and the first access mode does not discard the data. | Created with USE_HOST_PTR if a suitable host pointer is available, regardless of the first access mode. |
+| Host allocation command behaviour | Skipped if a suitable user host pointer is available. | In addition to skipping the allocation if a suitable user pointer is provided, the allocation is also skipped if the host command is created after its linked counterpart (it's retrieved via map operation instead). |
+| Memory transfer | Performed with read/write operations, non-host to non-host transfer is done with a host allocation as an intermediary (direct transfer is not supported by PI). | Only one allocation from the pair can be active at a time, the switch is done with map/unmap operations. |
```
Comment (bader) [inline sycl/doc/LinkedAllocations.md:62]: > Only one allocation from the pair can be active at a time, the switch is done with map/unmap operations.

Not sure I get how memory transfer between "non-host" devices is done in this case. Could you clarify, please?
Can we replace "non-host" with "device" (at least when we refer to allocations, memory, etc.)?

File: sycl/doc/LinkedAllocations.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+## Command linking approach
+
+Whenever two allocation commands are considered for linking, the decision is
+made based on the following criterion: the commands are linked if and only if
+the non-host device of the pair supports host unified memory (i. e. the device
```
Comment (bader) [inline sycl/doc/LinkedAllocations.md:68]:
```suggestion
the non-host device of the pair supports host unified memory (i.e. the device
```

File: sycl/doc/LinkedAllocations.md [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+and host share the same physical memory). The motivation for this is two-fold:
+- If the non-host device supports host unified memory, the USE_HOST_PTR flag
+should not result in any additional device memory allocation or copying between
+the two during map/unmap operations.
+- Even if the point above makes no difference for a particular pair of
+allocations (e. g. no host pointer is available for the non-host allocation),
```
Comment (bader) [inline sycl/doc/LinkedAllocations.md:74]:
```suggestion
allocations (e.g. no host pointer is available for the non-host allocation),
```
Usually abbreviations are not separated.

File: sycl/doc/LinkedAllocations.md (hunk 2 above)
Comment (sergey-semenov) [reply sycl/doc/LinkedAllocations.md]: >Not sure I get how memory transfer between "non-host" devices is done in this case. Could you clarify, please?

Added a sentence about the device-to-device case with a linked allocation.
> Can we replace "non-host" with "device" (at least when we refer to allocations, memory, etc.)?

Done.

Comment (bader) [reply sycl/doc/LinkedAllocations.md]: Thanks!

**Pull request merged**